### ISSUE TITLE
Moshi data classes not kept on minification

### ISSUE THREAD

**Issue by ektefa** [pusheco/pushe-android-studio-sample#71]
If you set minifyEnabled to true there will be an exception when receiving notifications.
So the fix would be adding `-keep class co.pushe.plus.** { *; }` to proguard-rules.
That might be helpful to mention this in the Readme.

Environement:
LG G8s Android 11
Pushe v2.6.3

**Comment by mahdi-malv** [comment]
Can you give some more details about the exception stack trace?

**Comment by ektefa** [comment]
```
2022-06-12 22:33:58.627 32024-32061/? E/Pushe: co.pushe.plus.messaging.MessageHandlingException: Unexpected error occurred on downstream message parsing
        at e3.s.a(R8$$SyntheticClass:20)
        at h9.t$b.g(ObservableFlatMap.java:2)
        at h9.s$a.g(ObservableFilter.java:3)
        at h9.f0$a.run(ObservableObserveOn.java:23)
        at a3.n.run(R8$$SyntheticClass:2)
        at k9.d$c$a.run(ExecutorScheduler.java:2)
        at k9.d$c.run(ExecutorScheduler.java:8)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.lang.NoSuchMethodException: co.pushe.plus.notification.messages.downstream.NotificationMessage.<init> [class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, class java.lang.String, interface java.util.List, interface g3.a, int, boolean, class java.lang.String, int, int, boolean, class java.lang.String, class java.lang.String, boolean, boolean, class java.lang.String, boolean, boolean, class java.lang.String, class java.lang.String, class java.lang.String, class p3.y, class java.lang.String, class java.lang.String, class java.util.Date, class java.lang.Long, class java.lang.Integer, interface java.util.Map, boolean, int, int, class kotlin.jvm.internal.DefaultConstructorMarker]
        at java.lang.Class.getConstructor0(Class.java:2332)
        at java.lang.Class.getDeclaredConstructor(Class.java:2170)
        at co.pushe.plus.notification.messages.downstream.NotificationMessageJsonAdapter.a(NotificationMessageJsonAdapter.kt:72)
        at com.squareup.moshi.JsonAdapter.c(JsonAdapter.java:2)
        at e3.s.a(R8$$SyntheticClass:8)
        at h9.t$b.g(ObservableFlatMap.java:2) 
        at h9.s$a.g(ObservableFilter.java:3) 
        at h9.f0$a.run(ObservableObserveOn.java:23) 
        at a3.n.run(R8$$SyntheticClass:2) 
        at k9.d$c$a.run(ExecutorScheduler.java:2) 
        at k9.d$c.run(ExecutorScheduler.java:8) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:923) 
```

**Comment by ektefa** [comment]
> Are you using `R8.fullMode`? If so, can you test again with that disabled?

I'm not. Just enabled minify.

**Comment by mahdi-malv** [comment]
Sorry I'm responding too late for this.
I expect Moshi data classes to be kept by moshi itself. Did a little research about it. Sounds like a bug from `Moshi:1.9.1`. However in latest version we use `1.12.0` which should not cause this bug. Are you using an old version of Pushe or `moshi` is old on your project?

**Comment by ektefa** [comment]
> Sorry I'm responding too late for this. I expect Moshi data classes to be kept by moshi itself. Did a little research about it. Sounds like a bug from [`Moshi:1.9.1`](https://stackoverflow.com/questions/58764077/moshi-1-9-1-throws-java-lang-nosuchmethodexception). However in latest version we use `1.12.0` which should not cause this bug. Are you using an old version of `moshi` on your project?

I'm using the latest pushe version I knew is there (2.6.3) and about moshi its not imported directly in the project. So that would be any version pushe is using. I also tried to keep moshi classes insted of pushe package itself and it didn't help.

**Comment by ektefa** [comment]
I'm using the latest pushe version I knew is there (2.6.3) and about moshi its not imported directly in the project. So that would be any version pushe is using. I also tried to keep moshi classes insted of pushe package itself and it didn't help.

> Sorry I'm responding too late for this. I expect Moshi data classes to be kept by moshi itself. Did a little research about it. Sounds like a bug from [`Moshi:1.9.1`](https://stackoverflow.com/questions/58764077/moshi-1-9-1-throws-java-lang-nosuchmethodexception). However in latest version we use `1.12.0` which should not cause this bug. Are you using an old version of `moshi` on your project?

I'm using the latest pushe version I knew is there (2.6.3) and about moshi its not imported directly in the project. So that would be any version pushe is using. I also tried to keep moshi classes insted of pushe package itself and it didn't help.

**Comment by mahdi-malv** [comment]
@ektefa Sorry for my delay reacting to this.
Still no idea why R8 removes `<init>` blocks out of classes. Please take a loot at this issue #75 
If it resolves the issue, You may continue the development or wait for the next bugFix.

**Comment by ektefa** [comment]
Tested `v2.6.4-alpha01`. Looks like issue is resolved in this version. Thanks.

**Comment by mahdi-malv** [comment]
Alright then. Stable release is also available (`2.6.4`). Though unchanged and just a bugfix
